plugin/transfer/transfer.go

```diff
@@ -0,0 +1,178 @@
+type Transferer interface {
+	// Transfer returns a channel to which it writes responses to the transfer request.
+	// If the plugin is not authoritative for the zone, it should immediately return the
+	// Transfer.ErrNotAuthoritative error.
+	//
+	// If serial is 0, handle as an AXFR request. Transfer should send all records
+	// in the zone to the channel. The SOA should be written to the channel first, followed
+	// by all other records, including all NS + glue records.
+	//
+	// If serial is not 0, handle as an IXFR request. If the serial is >
+	// the current serial for the zone, send a single SOA record to the channel.
+	// If the serial is >= the current serial for the zone, perform an AXFR fallback
+	// by proceeding as if an AXFR was requested (as above).
+	Transfer(zone string, serial uint32) (<-chan []dns.RR, error)
+}
+
+var (
+	// ErrNotAuthoritative is returned by Transfer() when the plugin is not authoritative for the zone
+	ErrNotAuthoritative = errors.New("not authoritative for zone")
+)
+
+// ServeDNS implements the plugin.Handler interface.
+func (t Transfer) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	if state.QType() != dns.TypeAXFR  && state.QType() != dns.TypeIXFR {
+		return plugin.NextOrFailure(t.Name(), t.Next, ctx, w, r)
+	}
+
+	// Find the first transfer instance for which the queried zone is a subdomain.
+	var x *xfr
+	for _, xfr := range t.xfrs {
+		zone := plugin.Zones(xfr.Zones).Matches(state.Name())
+		if zone == "" {
+			continue
+		}
+		x = xfr
+	}
+
+	// proceed with the queried zone
+	zone := state.Name()
+
+	if x == nil {
+		// Requested zone did not match any transfer instance zones.
+		// Pass request down chain in case later plugins are capable of handling transfer requests themselves.
+		return plugin.NextOrFailure(t.Name(), t.Next, ctx, w, r)
+	}
+
+	if !x.allowed(state) {
+		return dns.RcodeRefused, nil
+	}
+
+	// Get serial from request if this is an IXFR
+	var serial uint32
+	if state.QType() == dns.TypeIXFR {
+		soa, ok := r.Ns[0].(*dns.SOA)
+		if !ok {
+			return dns.RcodeServerFailure, nil
+		}
+		serial = soa.Serial
+	}
+
+	// Try each Transferer plugin
+	for _, p := range t.Transferers {
+		ch, err := p.Transfer(zone, serial)
+		if err == ErrNotAuthoritative {
+			// plugin was not authoritative for the zone, try next plugin
+			continue
+		}
+		if err != nil {
+			return dns.RcodeServerFailure, nil
+		}
+		x.send(state, ch)
+		for range ch {
+			// drain rest of channel if sendAxfr was aborted
+		}
+		return dns.RcodeSuccess, nil
+	}
+
+	return plugin.NextOrFailure(t.Name(), t.Next, ctx, w, r)
+
+}
+
+// send sends the response to the client.  It aborts if the first record received on the
+// channel is not an SOA.  send automatically adds the closing SOA to the end of an AXFR response.
+func (x xfr) send(state request.Request, fromPlugin <-chan []dns.RR) {
```

Comment:
the code is file/xfr.go is remarkbly shorter than this; why not use that?

Reply:
OK, the latest commit copies the code in from *file*, adapted to work with the channel returned by the plugin.

One functional difference I see is that *file*'s xfr implementation uses record count to cut/send envelopes, whereas previously I was using record data length. Not sure which one is more correct.

Another is that the code from *file* always sends the closing SOA in a separate envelope containing just that one record. Previously the closing SOA was included as the last record in the final envelope.

Not sure if these are significant.

Reply:
> Not sure if these are significant.

Are they?